### PR TITLE
chore: [IOPID-2579] add management of remote portal urls

### PR DESCRIPTION
**Pull request by Ladirico**
> [!Warning]
> This PR depends on https://github.com/pagopa/io-services-metadata/pull/916 https://github.com/pagopa/io-dev-api-server/pull/454 https://github.com/pagopa/io-app/pull/6579 https://github.com/pagopa/io-app/pull/6577

> [!Note]
> Need to edit `IO_SERVICES_METADATA_VERSION` ref on `scripts/generate-api-models.sh` after merge on `io-services-metadata`

## Short description
Add logic to manage the remote urls

## List of changes proposed in this pull request
- Add new ref  to metadata
- Add selector to get the value of remote urls from the cdn `absolutePortalLinksSelector`
- Add selector to manage the link building (base path from the cdn + body url) `replaceBaseUrlSelector`
- Edit hardcoded references to `https://io.italia.it/` and `https://ioapp.it/` using selectors
- Delete **unused** hardcoded references to `https://io.italia.it/` and `https://ioapp.it/` into the code 

## Demo 

https://github.com/user-attachments/assets/62888ce2-5e08-44b4-a521-e3c2c88b54a0

## How to test
Run the application and check that all links are working properly (as-is operation)

### CODE REVIEW

File: ts/urls.ts [hunk 1]
```diff
@@ -1,8 +1,12 @@
...
-export const acceptedPaymentMethodsFaqUrl =
-  "https://io.italia.it/metodi-pagamento";
+export const PRIVACY_URL_BODY = "app-content/tos_privacy.html";
+export const IO_SUPPLIERS_URL_BODY = "app-content/fornitori";
+export const UNSUPPORTED_DEVICE_FAQ_URL_BODY = "faq/#n1_15";
+
+// ITW urls
+export const ITW_IPZS_PRIVACY_URL_BODY = "informativa-ipzs";
+export const DOCUMENTS_ON_IO_FAQ_12_URL_BODY = "documenti-su-io/faq/#n1_12";
+export const DOCUMENTS_ON_IO_FAQ_14_URL_BODY = "documenti-su-io/faq/#n1_14";
```
Comment (ChrisMattew) [inline ts/urls.ts:12]: What about moving this variables in the `.env` file? What do you think?

Comment (Ladirico) [reply ts/urls.ts:12]: These are constant values that do not change depending on the environment (they are the same in production and in development), so for best practice they should not be included in .env files

Comment (ChrisMattew) [reply ts/urls.ts:12]: It makes sense!

File: ts/config.ts [hunk 2]
```diff
@@ -150,30 +150,12 @@ export const totMessageFetchWorkers = pipe(
 export const shufflePinPadOnPayment =
   Config.SHUFFLE_PINPAD_ON_PAYMENT === "YES";
 
-export const privacyUrl: string = pipe(
-  Config.PRIVACY_URL,
-  t.string.decode,
-  E.getOrElse(() => "https://io.italia.it/app-content/tos_privacy.html")
-);
-
```
Comment (shadowsheep1) [inline ts/config.ts:158]: There is an ESLint rules about this `privacyUrl` constant. Maybe we don't need it anymore. What do you think?

File: ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts [hunk 3]
```diff
@@ -1,56 +0,0 @@
-export type TmpDiscountType = {
-  category: string;
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts:2]: Why did we delete this mock? 🤔

File: ts/features/itwallet/walletInstance/hook/useItwWalletInstanceRevocationAlert.ts [hunk 4]
```diff
@@ -15,31 +20,53 @@ const alertCtaText = I18n.t(
...
+    ])
   );
 };
 
 /**
  * Displays an alert based on the revocation reason.
  */
 const showWalletRevocationAlert = (
+  itwMinIntegrityReqUrl: string,
+  itwDocsOnIOMultipleDevicesUrl: string,
```
Comment (shadowsheep1) [inline ts/features/itwallet/walletInstance/hook/useItwWalletInstanceRevocationAlert.ts:69]: Where did we use this new parameters inside the function body? I cannot see changes in the function body. Did I miss something?

File: ts/store/reducers/__mock__/backendStatus.ts [hunk 5]
```diff
@@ -303,6 +308,7 @@ export const baseRawBackendStatus: BackendStatus = {
       },
       ipatente_cta_visible: true,
       feedback_banner_visible: true,
+      ipatente_cta_visible: true,
```
Comment (shadowsheep1) [inline ts/store/reducers/__mock__/backendStatus.ts:311]: Watch out about the warning. Here we have two `ipatente_cta_visible`.

File: ts/store/reducers/__mock__/backendStatus.ts [hunk 6]
```diff
@@ -396,6 +406,7 @@ export const baseBackendConfig: Config = {
     },
     ipatente_cta_visible: true,
     feedback_banner_visible: true,
+    ipatente_cta_visible: true,
```
Comment (shadowsheep1) [inline ts/store/reducers/__mock__/backendStatus.ts:409]: As above

File: ts/store/reducers/backendStatus/remoteConfig.ts [hunk 7]
```diff
@@ -296,6 +296,51 @@ export const isIdPayEnabledSelector = createSelector(
...
+    )
+);
+
+type hostType = keyof ReturnType<typeof absolutePortalLinksSelector>;
+
+/**
+ * Selector to generate a complete URL by replacing the base domain with a dynamic value from absolutePortalLinks.
+ * This is useful when URLs need to be dynamically constructed based on remote configuration or state.
+ */
+export const replaceBaseUrlSelector = createSelector(
```
Comment (shadowsheep1) [inline ts/store/reducers/backendStatus/remoteConfig.ts:323]: Actually here we don't replace anything, but we add the base url, am I right? What about changing the function name to something like `addBaseUrlSelector`. What about explicitly state that we add also the schema along with the base host? What do you think?

File: ts/store/reducers/backendStatus/remoteConfig.ts [hunk 8]
```diff
@@ -296,6 +296,51 @@ export const isIdPayEnabledSelector = createSelector(
...
+  (_: GlobalState, baseUrl: hostType) => baseUrl,
+
+  // Step 3: Extract the body URL (the path to append to the base URL) from the input parameters.
+  (_: GlobalState, __: hostType, bodyUrl: string) => bodyUrl,
+
+  // Step 4: Combine the absolutePortalLinks object, the base URL key, and the body URL to create the full URL.
+  (absolutePortalLinks, baseUrl, bodyUrl) => {
+    try {
+      // Concatenate the base URL with the body URL.
+      return `${absolutePortalLinks[baseUrl]}${bodyUrl}`;
```
Comment (shadowsheep1) [inline ts/store/reducers/backendStatus/remoteConfig.ts:337]: Here we are not dealing with trailing slashes `/`. What if we get from the remote config `https://www.example.com` instead of `https://www.example.com/`? What do you think?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts (hunk 3 above)
Comment (Ladirico) [reply ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts:2]: When I was looking for references to io.italia.it, the reference to this mock file came up and I saw that it was not being used, so I deleted it. I can also reinsert it if you think it necessary

File: ts/features/itwallet/walletInstance/hook/useItwWalletInstanceRevocationAlert.ts (hunk 4 above)
Comment (Ladirico) [reply ts/features/itwallet/walletInstance/hook/useItwWalletInstanceRevocationAlert.ts:69]: You do not see any changes in the body of the function because these parameters are named like the variables that were there before. If you pay attention to the code, you will see that before there were two constants that contained links to the URL you needed. The logic I created generates the link using a selector to which you pass parameters. In this particular case to leave the UI component without logic I passed the URL parameter as prop

File: ts/store/reducers/backendStatus/remoteConfig.ts (hunk 8 above)
Comment (Ladirico) [reply ts/store/reducers/backendStatus/remoteConfig.ts:337]: Rather than adding a configuration/rule here I would add a rule in io-services-metadata. What do you think?

File: ts/store/reducers/__mock__/backendStatus.ts (hunk 5 above)
Comment (Ladirico) [reply ts/store/reducers/__mock__/backendStatus.ts]: Yes! Correct! Thank you for noticing. Unfortunately, I had to add that value in order not to make the configurations go wrong because a PR I was dependent on was missing from merging. Anyway now the PR is merged and i've delete this!

File: ts/config.ts (hunk 2 above)
Comment (shadowsheep1) [reply ts/config.ts:158]: We have this [ESLint rule](https://github.com/pagopa/io-app/blob/f2659cfd53feff5976397807916b84e234287d04/.eslintrc.js#L141). I guess we don't need it anymore. What do you think?

File: ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts (hunk 3 above)
Comment (shadowsheep1) [reply ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts:2]: It's okay to remove, then. I'd just double check with the CGN Team if we didn't already.

File: ts/features/itwallet/walletInstance/hook/useItwWalletInstanceRevocationAlert.ts (hunk 4 above)
Comment (shadowsheep1) [reply ts/features/itwallet/walletInstance/hook/useItwWalletInstanceRevocationAlert.ts:69]: Okay perfect then!

File: ts/store/reducers/backendStatus/remoteConfig.ts (hunk 8 above)
Comment (shadowsheep1) [reply ts/store/reducers/backendStatus/remoteConfig.ts:337]: A good rule of thumb is to sanitize data on both sides of `producer` and `consumer`. So I'd suggest to add both. What do you think?

Comment (shadowsheep1) [reply ts/store/reducers/backendStatus/remoteConfig.ts:337]: I'd suggest to add a deep test suite to check the correct behavior of this function and to avoid edge cases. What do you think?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts (hunk 3 above)
Comment (Ladirico) [reply ts/features/bonus/cgn/__mock__/availableMerchantDetail.ts:2]: Done, for CGN team is safe remove this mock!

File: ts/features/tos/store/selectors/index.ts [hunk 9]
```diff
@@ -2,24 +2,27 @@ import { createSelector } from "reselect";
...
+import { PRIVACY_URL_BODY } from "../../../../urls";
 
 export const tosConfigSelector = createSelector(
   remoteConfigSelector,
-  remoteConfig =>
-    pipe(
+  absolutePortalLinksSelector,
+  (remoteConfig, absolutePortalLinks) => {
+    const DEFAULT_TOS_CONFIG: TosConfig = {
+      tos_url: `${absolutePortalLinks.io_showcase}${PRIVACY_URL_BODY}`,
```
Comment (shadowsheep1) [inline ts/features/tos/store/selectors/index.ts:16]: Just thinking about. If for some reason we get this value, before the privacy url has been ported to the new showcase website, we cannot navigate the url, showing the citizen an error. Am I right? How can we mitigate this remote scenario? Can we think about adding a temporary redirect  on` ioapp.it` so that if someone try to navigate `https://ioapp.it/app-content/tos_privacy.html` it'll be redirected to `https://io.italia.it/app-content/tos_privacy.html`? We can write this note down and try to address it the next sprint on `ioapp.it`. 
What do you think?

Comment (Ladirico) [reply ts/features/tos/store/selectors/index.ts:16]: I invite you to check the logic of the ‘absolutePortalLinksSelector’. In the event that the remote value is not available, a harcoded fallback value is taken in the code https://github.com/pagopa/io-app/pull/6594/files#diff-c235798ce73b51bc5997e2e6e8091267d1392b740b63fb5767af63943a8ae959.

File: ts/store/reducers/backendStatus/remoteConfig.ts (hunk 8 above)
Comment (Ladirico) [reply ts/store/reducers/backendStatus/remoteConfig.ts]: done!

File: ts/features/tos/store/selectors/index.ts (hunk 9 above)
Comment (Ladirico) [reply ts/features/tos/store/selectors/index.ts:16]: Good point, but for the time now (until the switch takes place) the links will point to existing portals, not to new ones, so the problem you mention does not exist. 
I should also point out that the absolutePortalLinksSelector has a local fallback that is used when URLs cannot be retrieved from io-services-metadata. Currently the fallback refers to as-is links

Comment (shadowsheep1) [reply ts/features/tos/store/selectors/index.ts:16]: Gotcha! You're right!

**Review by shadowsheep1 — APPROVED** [review]
LGTM

**Comment by shadowsheep1** [comment]
> [!warning]
> Remember to merge the PR on [io-services-metadata](https://github.com/pagopa/io-services-metadata/pull/916) and create the new version tag. Then change that: `IO_SERVICES_METADATA_VERSION=IOPID-2581-add-portal-links` before merging.
